tensorflow/python/distribute/multi_process_runner.py

```diff
@@ -988,16 +994,72 @@ def run(self, fn, args=None, kwargs=None):
       conn.send((fn, args or [], kwargs or {}))
 
     process_statuses = []
+    total_timeout = 300  # 5 minutes total timeout for all processes
+    start_time = time.time()
+    
     for (task_type, task_id), conn in self._conn.items():
-      logging.info('Waiting for the result from %s-%d', task_type, task_id)
-      try:
-        process_statuses.append(conn.recv())
-      except EOFError:
-        # This shouldn't happen due to exceptions in fn. This usually
-        # means bugs in the runner.
-        self.shutdown()
-        raise RuntimeError('Unexpected EOF. Worker process may have died. '
-                           'Please report a bug')
+      remaining_time = total_timeout - (time.time() - start_time)
+      if remaining_time <= 0:
+        raise RuntimeError('Total timeout of %d s '
+                           'exceeded while waiting for processes',
+                            total_timeout)
```

Comment:
For exceptions you have to use `"str %d" % arg"` (note the extra `%`, not separate by comma), or, better, f-string formatting. The exception argument is not evaluated lazily, so there's no need to not use f-strings.